jstz_api/src/encoding/text_decoder.rs

```diff
@@ -0,0 +1,381 @@
+// };
+
+// [Exposed=*]
+// interface TextDecoder {
+//   constructor(optional DOMString label = "utf-8", optional TextDecoderOptions Options = {});
+
+//   USVString decode(optional AllowSharedBufferSource input, optional TextDecodeOptions Options = {});
+// };
+// TextDecoder includes TextDecoderCommon;
+
+// https://encoding.spec.whatwg.org/#interface-mixin-textdecodercommon
+// interface mixin TextDecoderCommon {
+//   readonly attribute DOMString encoding;
+//   readonly attribute boolean fatal;
+//   readonly attribute boolean ignoreBOM;
+// };
+
+#[derive(Trace, Finalize)]
+pub struct TextDecoder {
+    encoding: &'static Encoding,
+    #[unsafe_ignore_trace]
+    decoder: Decoder,
+    io_queue: Vec<u8>,
+    ignore_bom: bool,
+    error_mode: String,
+    do_not_flush: bool,
+}
+#[derive(Trace, Finalize, Default)]
+pub struct TextDecoderOptions {
+    fatal: bool,
+    ignore_bom: bool,
+}
+
+impl TryFromJs for TextDecoderOptions {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        let obj = value.as_object().ok_or_else(|| {
+            JsError::from_native(JsNativeError::typ().with_message("Expected `JsObject`"))
+        })?;
+        let def = TextDecoderOptions::default();
+        let fatal = if obj.has_property(js_string!("fatal"), context)? {
+            obj.get(js_string!("fatal"), context)?
+                .try_js_into(context)?
+        } else {
+            def.fatal
+        };
+        let ignore_bom = if obj.has_property(js_string!("ignoreBOM"), context)? {
+            obj.get(js_string!("ignoreBOM"), context)?
+                .try_js_into(context)?
+        } else {
+            def.ignore_bom
+        };
+        Ok(Self { fatal, ignore_bom })
+    }
+}
+
+#[derive(Trace, Finalize, Default)]
+pub struct TextDecodeOptions {
+    stream: bool,
+}
+
+impl TryFromJs for TextDecodeOptions {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        let obj = value.as_object().ok_or_else(|| {
+            JsError::from_native(JsNativeError::typ().with_message("Expected `JsObject`"))
+        })?;
+        let def = TextDecodeOptions::default();
+        let stream = if obj.has_property(js_string!("stream"), context)? {
+            obj.get(js_string!("stream"), context)?
+                .try_js_into(context)?
+        } else {
+            def.stream
+        };
+        Ok(Self { stream })
+    }
+}
+
+impl TextDecoder {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message(
+                        "Failed to convert js value into rust type `TextDecoder`",
+                    )
+                    .into()
+            })
+    }
+
+    //  https://encoding.spec.whatwg.org/#dom-textdecoder
+    fn constructor(
```

Comment:
Convention: `new` instead of `constructor`

Reply:
done